### ISSUE TITLE
External script doesn't always load when added to Nuxt page via "head()"

### ISSUE THREAD

**Issue by markplewis** [nuxt/nuxt.js#5052]
### Version

[v2.4.3](https://github.com/nuxt.js/releases/tag/v2.4.3)

### Reproduction link

[https://github.com/markplewis/nuxt-external-resource-loading-issue](https://github.com/markplewis/nuxt-external-resource-loading-issue)

### Steps to reproduce

Clone [my demo repo](https://github.com/markplewis/nuxt-external-resource-loading-issue) and then do the following:

## Case 1

1. Launch the app via `npm run dev` and open it up in your browser.
2. Click the "Test 1" link.
3. The following error will appear in your console:
```
TypeError: window.jQuery is not a function
  at VueComponent.mounted (test-1.js:34)
```
4. Refresh the page.
5. Everything will load correctly.
6. Click the "Home" link.
7. Click the "Test 1" link.
8. Everything will load correctly.
9. Click the "Home" link.
10. Refresh the page.
11. Click the "Test 1" link.
12. The following error will appear in your console:
```
TypeError: window.jQuery is not a function
  at VueComponent.mounted (test-1.js:34)
```
13. Refresh the page.
14. Everything will load correctly.

## Case 2

1. Launch the app via `npm run dev` and open it up in your browser.
2. Click the "Test 2" link.
3. The "(CDN script has loaded)" message will **not** be appended to the `<h1>` element.
4. Refresh the page.
5. The "(CDN script has loaded)" message will appear inside the `<h1>`.
6. Click the "Home" link.
7. Click the "Test 2" link.
8. The "(CDN script has loaded)" message will appear inside the `<h1>`.
9. Click the "Home" link.
10. Refresh the page.
11. Click the "Test 2" link.
12. The "(CDN script has loaded)" message will **not** be appended to the `<h1>` element.
13. Refresh the page.
14. The "(CDN script has loaded)" message will appear inside the `<h1>`.

### What is expected ?

I'm trying to load an external script into a Nuxt page via the page's `head`:
```
head() {
  return {
    script: [{
      src: "https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.slim.min.js"
    }]
  };
}
```
I expect the script to load and be available every time the page loads, regardless of whether it was served directly from the server or rendered on the client-side.

### What is actually happening?

Unfortunately, the external script doesn't seem to load when the page is server-rendered - it only seems to work when the page is client-rendered. I discovered the following issues, but neither of them have helped me solve this problem:

- https://github.com/nuxt/nuxt.js/issues/3396
- https://github.com/nuxt/nuxt.js/issues/3872

In that first issue, manniL said the following:

> "Loading external scripts through the head function works fine for me. I'd suggest to do this on a `layout` basis anyway. For pages there is the caveat that you can't ensure the readiness easily."

What does "you can't ensure the readiness easily" mean? I'd prefer not to move this script into my layout because it will only be needed for one particular page, and I don't want to force users to download unnecessary resources.

I can avoid errors (see test case 1, above) by wrapping my calls to `window.jQuery`, like this: 
```
mounted() {
  if (!process.server && window.jQuery) {
    ...  
  }
}
```
But this leaves me with the problem that I described in test case 2 (see above).

<div align="right"><sub><em>This bug report is available on <a href="https://cmty.app/nuxt">Nuxt</a> community (<a href="https://cmty.app/nuxt/nuxt.js/issues/c8677">#c8677</a>)</em></sub></div>

**Comment by chanmathew** [comment]
I also faced this issue trying to load the Stripe API (v3) in a page, often it'll say Stripe library isn't loaded.

**Comment by markplewis** [comment]
Apologies for not replying in time to prevent the bot from closing this issue. I just updated [my test repo](https://github.com/markplewis/nuxt-external-resource-loading-issue) to Nuxt 2.6.1 and I'm still experiencing the problems described above.

**Comment by dysol** [comment]
I'm having similar issue when trying to load Youtube API on one particular page.

I've also tried using npm vue-script2 to have more control over external scripts loading.

`
VueScript2.load('https://www.youtube.com/iframe_api').then(() => {
    console.log(new YT.Player('ytplayer')); (YT is the global object inserted from the API)
});
`

getting the below error
`Uncaught (in promise) TypeError: YT.Player is not a constructor`

**Comment by abh153k** [comment]
Does anyone have a solution to this problem?

**Comment by DidelotK** [comment]
I have the same problem with Stripe and Google maps :p, anyone have a solution to this problem ?
Is it due to nuxt specific version ?

Nuxt version: 2.6.2

Example to reproduce: `pages/toto.vue`
`      script: [
        { src: 'https://js.stripe.com/v3/', defer: true },
        {
          src: `https://maps.googleapis.com/maps/api/js?key=${googleApiKey}&libraries=places`
        }
      ],`

**Comment by abh153k** [comment]
@DidelotK All versions are affected, AFAIK.

**Comment by rchl** [comment]
@markplewis and others: It all works as expected and has to do with how html/JS works, not Nuxt.

In `Case 1`, you are navigating in SPA fashion to another page and that adds a script and you expect it to be loaded in `mounted` hook already. That's false assumption as scripts appended dynamically don't load in sync fashion. You have to listen to `load` event to know when they have loaded and only then you can start using code that it defines.

In `Case 2` is a slight variation but pretty much same reason for the behavior. It works on reloading the page as then you'll get the script appended already on the server so it will load synchronously and you'll have jQuery loaded by the time `mounted` is fired.

Basically you have to understand the difference between script added programatically (through DOM) and script that is in HTML of the page already. Former will load asynchronously while latter will load synchronously.

So to make this work consistently, rather than using `head` function, you need to use something like `VueScript2` to load script dynamically and wait for it to load. For jQuery this is enough. For other APIs like YouTube you might also need to wait for the API code to tell you that it has loaded because it might have some async initialisation routine that has to run even after script has loaded (refer to APIs documentation how to handle that).

**Comment by markplewis** [comment]
Thanks @rchl. This is the simplest example that I could come up with and it seems to be working:
```javascript
export default {
  mounted() {
    if (!process.server && !window.jQuery) {
      const script = document.createElement("script");
      script.onload = this.onScriptLoaded;
      script.type = "text/javascript";
      script.src = "https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.slim.min.js";
      document.head.appendChild(script);
    } else {
      this.onScriptLoaded();
    }
  },
  methods: {
    onScriptLoaded(event = null) {
      if (event) {
        console.log("Was added");
      } else {
        console.log("Already existed");
      }
      console.log(window.jQuery);
      window.jQuery("h1").append(` <span>(CDN script has loaded)</span>`);
    }
  }
}
```

**Comment by rchl** [comment]
Looks good to me. Something like `VueScript2` would abstract that and make it look a bit nicer IMO but this is nice, no-lib solution.

BTW. `mounted` is only executed on the client so check for `!process.server` is not needed.

**Comment by webcoderkz** [comment]
It's version 2.10 and i still experience this issue, lol. What's the status?

**Comment by hung5s** [comment]
I created this utility function to load external JS file as Nuxt page's head property doesn't work. In my case, It only adds <script> to <head> if the page is requested directly, not as a route-view component.

` loadJsFile(url, id, onLoadedCallback, defer, async) {
    if (process.browser) {
      let test = document.getElementById(id);
      if (test) return onLoadedCallback();
      else {
        const script = document.createElement("script")
        script.src = url
        script.id = id
        script.onload = onLoadedCallback
        script.type = "text/javascript"
        script.defer = defer == true || defer == undefined ? true : false
        script.async = async == true || defer == undefined ? true : false
        document.head.appendChild(script)
      }
    }
  }`

**Comment by manniL** [comment]
Thanks to vue-meta, this should be way easier now. I've wrote a bit about my preferred way in [this article](https://vueschool.io/articles/vuejs-tutorials/how-to-load-third-party-scripts-in-nuxt-js/).

**Comment by broox** [comment]
i'm attempting to use VueMeta to inject an external library into a component and am experiencing an issue where, in some cases, while navigating between pages, the `script.callback()` function is not called. is this expected if the JS is cached? or should this callback be fired no matter what?

```
export default {
  head () {
    return {
      script: [
        {
          hid: 'mapbox',
          src: 'https://api.mapbox.com/mapbox-gl-js/v2.6.1/mapbox-gl.js',
          defer: true,
          callback: () => { console.log('Map script loaded'); this.render() }
        }
      ]
    }
  },
  ...
}
```

it seems to only happen when navigating from another "SPA mode" page that has this component on it... but not in all cases.

**Comment by jonathan-ingram** [comment]
> i'm attempting to use VueMeta to inject an external library into a component and am experiencing an issue where, in some cases, while navigating between pages, the `script.callback()` function is not called. is this expected if the JS is cached? or should this callback be fired no matter what?
> 
> ```
> export default {
>   head () {
>     return {
>       script: [
>         {
>           hid: 'mapbox',
>           src: 'https://api.mapbox.com/mapbox-gl-js/v2.6.1/mapbox-gl.js',
>           defer: true,
>           callback: () => { console.log('Map script loaded'); this.render() }
>         }
>       ]
>     }
>   },
>   ...
> }
> ```
> 
> it seems to only happen when navigating from another "SPA mode" page that has this component on it... but not in all cases.

Hello @broox Did you ever locate a solution to this? I only ask because I'm experiencing this random behaviour too. At first I could only recreate this bug on Safari, but have now concluded it's browser-agnostic.

For my environment it seems to only occur once in roughly 50 page loads. And when it does occur it's always when transitioning between two pages who both refer to the same 3rd-party JavaScript file. And it's the same outcome that you're experiencing (in that the `callback` function never fires).

Currently using **Nuxt v2.15.8**

**Comment by simoroma** [comment]
I would this work with `<script setup>`?
